x/aswap/codec.proto

```diff
@@ -0,0 +1,67 @@
+syntax = "proto3";
+
+package aswap;
+
+import "github.com/iov-one/weave/codec.proto";
+import "github.com/iov-one/weave/coin/codec.proto";
+
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+
+// Swap is designed to hold some coins for atomic swap, locked by preimage_hash
+message Swap {
+    // metadata is used for schema versioning support
+    weave.Metadata metadata = 1;
+
+    bytes src = 2 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+    // preimage hash is a hash of user's secret preimage that would serve as a "password"
+    // to unlock the coins in the swap should the transaction be a success
+    bytes preimage_hash = 3;
+    bytes recipient = 4 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+
+    // If unreleased before timeout, swap will return coins to src.
+    // Timeout represents wall clock time as read from the block header. Timeout
+    // is represented using POSIX time format.
+    // Expiration time is inclusive meaning that the swap expires as soon as
+    // the current time is equal or greater than timeout value.
+    // nonexpired: [created, timeout)
+    // expired: [timeout, infinity)
+    int64 timeout = 5 [(gogoproto.casttype) = "github.com/iov-one/weave.UnixTime"];
+
+    // max length 128 characters
+    string memo = 6;
+}
+
+// CreateSwapMsg creates a Swap with some coins.
+message CreateSwapMsg {
+    weave.Metadata metadata = 1;
+
+    bytes src = 2 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+    // sha256 hash of preimage
+    bytes preimage_hash = 3;
+    bytes recipient = 4 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+
+    // amount may contain multiple token types
+    repeated coin.Coin amount = 5;
+    // Timeout represents wall clock time.
+    int64 timeout = 6 [(gogoproto.casttype) = "github.com/iov-one/weave.UnixTime"];
+    // max length 128 character
+    string memo = 7;
+}
+
+// ReleaseSwapMsg releases the tokens to the recipient.
+// This operation is authorized by preimage, which is sent raw and then hashed on the backend.
+message ReleaseSwapMsg {
+    weave.Metadata metadata = 1;
+
+    bytes swap_id = 2 [(gogoproto.customname) = "SwapID"];
```

Comment:
q: I wonder if the `swapID` is adding too much complexity to this. The second party who is going to claim the coins will know the preimage_hash from the sender and the preimage itself from the other chain. The `swapID` is something new on our chain only and maybe not easily available to the second party.
If we don't need to support multiple swaps with the same preimage we can simply find the swap by the preimage hash.

💡 Thinking this a bit further the `preimage_hash` may be used as the key in the bucket so that we have a unique constraint and an index out of the box. I would still generate a condition address with that value and store it with the swap to point to the coins. This would simplify access and may be generic enough to reuse functionality with escrow later.

Reply:
I was thinking this:
1. Still have swapID for Address and internal sequencing
2. Remove swapID from messages, fetch by preimage_hash
3. Profit.

I'll redo this in a bit more elegant way.
Also, I keep wondering if we need indices for src/recipient at all, maybe we should only have a key by preimage_hash and that's it?

Reply:
@alpe done, feel free to take a look when you have a moment. Looks cleaner now

Reply:
Actually, we can go even further and use preimage_hash as the base for the condition -> address. I'm going to introduce that now too

Reply:
And simplifying it even further we don't even need to store SwapAddress, but instead generate it out of the PreimageHash on-demand!

Reply:
Also, there's a possibility to improve performance even further by just truncating PreimageHash to address size, without hashing it again, but I'm not sure it's such a good idea.